cmd/admin-prometheus-metrics-v3.go

```diff
@@ -0,0 +1,203 @@
+	subsys := req.subsystem
+	switch subsys {
+	case "", "api", "system", "debug", "cluster", "ilm", "audit", "logger", "replication", "notification", "scanner":
+	default:
+		fatalIf(errInvalidArgument().Trace(), "invalid metric type `"+subsys+"`")
+	}
+
+	list := ctx.Bool("list")
+	bucket := ctx.String("bucket")
+	metricsURL := req.aliasURL + metricsV3EndPointRoot
+	params := url.Values{}
+
+	if len(bucket) > 0 {
+		bms := strings.Join(bucketMetricsSubSystems.ToSlice(), ", ")
+		if len(subsys) == 0 {
+			fatalIf(errInvalidArgument().Trace(), fmt.Sprintf("metric type must be passed with --bucket. valid values are `"+bms+"`"))
+		}
+
+		if !bucketMetricsSubSystems.Contains(subsys) {
+			fatalIf(errInvalidArgument().Trace(), fmt.Sprintf("--bucket is applicable only for metric types `"+bms+"`"))
+		}
+
+		metricsURL += "/bucket"
+	}
+
+	if len(subsys) > 0 {
+		metricsURL += "/" + subsys
+	}
+
+	if len(bucket) > 0 {
+		// bucket specific metrics endpoints have '/bucket' prefix and bucket name as suffix.
+		// e.g. bucket api metrics: /bucket/api/mybucket
+		metricsURL += "/" + bucket
+	}
+
+	if list {
+		params.Add("list", "true")
+	}
+
+	qparams := params.Encode()
+	if len(qparams) > 0 {
+		metricsURL += "?" + qparams
+	}
+
+	resp, e := fetchMetrics(metricsURL, req.token)
+	if e != nil {
+		return e
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		printMsg(prometheusMetricsV3Reader{
+			prometheusMetricsReader: prometheusMetricsReader{Reader: resp.Body},
+			isList:                  list,
+		})
+		return nil
+	}
+
+	return errors.New(resp.Status)
+}
+
+// JSON returns jsonified message
+func (pm prometheusMetricsV3Reader) JSON() string {
+	if !pm.isList {
+		return pm.prometheusMetricsReader.JSON()
+	}
+
+	metricInfos := []prometheusMetricInfo{}
+	rows := pm.readMetricsV3List()
+	for idx, row := range rows {
+		if idx == 0 || len(row) != 4 {
+			// first row is the header, skip it.
+			continue
+		}
+		metricInfos = append(metricInfos, prometheusMetricInfo{
+			Name:   strings.TrimSpace(row[0]),
+			Type:   strings.TrimSpace(row[1]),
+			Help:   strings.TrimSpace(row[2]),
+			Labels: strings.TrimSpace(row[3]),
+		})
+	}
+
+	jsonMessageBytes, e := json.MarshalIndent(metricInfos, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+	return string(jsonMessageBytes)
+}
+
+func (pm prometheusMetricsV3Reader) readMetricsV3List() [][]string {
+	scanner := bufio.NewScanner(pm.Reader)
+	lineNum := 0
+	rows := [][]string{}
+	for scanner.Scan() {
+		line := scanner.Text()
+		lineNum++
+		if lineNum == 2 {
+			// second line is just a separator, skip it.
+			continue
+		}
+		// remove leading and trailing '|', remove '`' and finally split by '|'
+		// example row: | `minio_bucket_api_total` | `counter` | Total number of requests for a bucket | `bucket,name,type,server` |
+		row := strings.Split(strings.ReplaceAll(strings.Trim(line, "|"), "`", ""), "|")
```

Comment:
can you show me a sample line here and why are we doing this manually like this?

Reply:
Sample line is there in the comment @harshavardhana 

The full api response is a markdown table e.g.
```
| Name | Type | Help | Labels |
| ---- | ---- | ---- | ------ |
| `minio_bucket_api_traffic_received_bytes` | `counter` | Total number of bytes sent for a bucket | `bucket,type,server` |
| `minio_bucket_api_traffic_sent_bytes` | `counter` | Total number of bytes received for a bucket | `bucket,type,server` |
| `minio_bucket_api_inflight_total` | `gauge` | Total number of requests currently in flight for a bucket | `bucket,name,type,server` |
| `minio_bucket_api_total` | `counter` | Total number of requests for a bucket | `bucket,name,type,server` |
| `minio_bucket_api_canceled_total` | `counter` | Total number of requests canceled by the client for a bucket | `bucket,name,type,server` |
| `minio_bucket_api_4xx_errors_total` | `counter` | Total number of requests with 4xx errors for a bucket | `bucket,name,type,server` |
| `minio_bucket_api_5xx_errors_total` | `counter` | Total number of requests with 5xx errors for a bucket | `bucket,name,type,server` |
| `minio_bucket_api_ttfb_seconds_distribution` | `counter` | Distribution of time to first byte across API calls for a bucket | `bucket,name,le,type,server` |
```

I'm trying to extract this data so that we can either print it as a nice table on terminal or return in proper JSON format when `--json` is passed.

Reply:
so you are relying on the fact that this markdown will be kept in-tact in this manner?

Reply:
as discussed, will remove this (`--list`) for now, as a better implementation will require changes on the server side.